### PR TITLE
Question about rewirings in parsed-program-to-logical-matrix

### PR DESCRIPTION
**Pull request by appleby**
While gazing at these lovely codes I noticed that in `quil::parsed-program-to-logical-matrix`, when there is a combined entering/exiting rewiring on a single instruction, it calls the locally-defined function `apply-entering-rewiring` on the `exiting-vector` and `apply-exiting-rewiring` on the `entering-vector`. This seemed backwards to me, but I am noob and lack understanding.

Is this a bug? Expected?

I wrote a test that I think indicates a bug, but I might be holding it wrong. Prior to these changes, the new test fails like so (after, all tests pass):

``` common-lisp
CL-USER> (fiasco:run-tests 'cl-quil-tests::test-parsed-program-to-logical-matrix-entering-exiting-rewiring)
TEST-PARSED-PROGRAM-TO-LOGICAL-MATRIX-ENTERING-EXITING-REWIRING...........[FAIL]


Test run had 1 failure:

  Failure 1: FAILED-ASSERTION when running CL-QUIL-TESTS::TEST-PARSED-PROGRAM-TO-LOGICAL-MATRIX-ENTERING-EXITING-REWIRING
    Binary predicate (OPERATOR= X Y) failed.
    x: (CL-QUIL:PARSED-PROGRAM-TO-LOGICAL-MATRIX CL-QUIL-TESTS::PP) => #<MAGICL:MATRIX [Z] 4x4:
                                                                          1.000 + 0.000j     0.000 + 0.000j     0.000 + 0.000j     0.000 + 0.000j
                                                                          0.000 + 0.000j     0.000 + 0.000j     1.000 + 0.000j     0.000 + 0.000j
                                                                          0.000 + 0.000j     0.000 + 0.000j     0.000 + 0.000j     1.000 + 0.000j
                                                                          0.000 + 0.000j     1.000 + 0.000j     0.000 + 0.000j     0.000 + 0.000j>
    y: (CL-QUIL:PARSED-PROGRAM-TO-LOGICAL-MATRIX CL-QUIL-TESTS::PP-REWIRED) => #<MAGICL:MATRIX [Z] 4x4:
                                                                                  1.000 + 0.000j     0.000 + 0.000j     0.000 + 0.000j     0.000 + 0.000j
                                                                                  0.000 + 0.000j     0.000 + 0.000j     0.000 + 0.000j     1.000 + 0.000j
                                                                                  0.000 + 0.000j     1.000 + 0.000j     0.000 + 0.000j     0.000 + 0.000j
                                                                                  0.000 + 0.000j     0.000 + 0.000j     1.000 + 0.000j     0.000 + 0.000j>
NIL
(#<test-run of TEST-PARSED-PROGRAM-TO-LOGICAL-MATRIX-ENTERING-EXITING-REWIRING: 1 test, 1 assertion, 1 failure in 0.0 sec (1 failed assertion, 0 errors, none expected)>)
```

### CODE REVIEW

**Comment by ecpeterson** [comment]
This is indeed a bug! I know how this one came about: I originally destructured the payload into `(exiting-vector . entering-vector)`, which had the correct behavior because I evidently mixed the behavior up in two places. I noticed this first error, fixed it, then didn't re-test (or re-read) the rest of the function. Good eye.

**Comment by appleby** [comment]
> p.s. "lovely codes" is flattery that quilc hasn't completely earned, but thank you all the same

Story time!

A couple years ago, I decided to try painting. I thought it would be a fun hobby and way to cheaply fill my house with art I enjoyed. Unfortunately, while I do enjoy the *process* of painting, I rarely enjoy the result. Partially this is because I am a novice painter and my paintings reflect that. But partially, the mild disgust I feel for my paintings is a natural result of the way a painting is made, namely: 1) start with a rough sketch or idea, 2) proceed by iterative refinement removing the worst of the errors and thereby asymptotically approaching the idea in your head for the how the painting should look. But in the process of fixing one error, you introduce a new one. And so the process goes, around and around, fixing and introducing warts. But always focused on the warts. As a result, a paining is never really "finished", just abandoned (often in despair) when you finally accept that any further changes are unlikely to improve it. So when I look at my paintings, what I see is not the lovely bits (if there are any), but the warts that escaped my grasp.

Now as you move along the gradient from novice -> expert you become more skilled and introduce fewer warts. But the keen wart-detecting eye you developed as a novice probably stays with you forever.

tl;dr

The moral of the story? The "artist" rarely finds beauty in their work. But the well-intentioned observer rarely finds fault.

Thanks for coming to my TED talk!
